src/NGen.cpp

```diff
@@ -539,6 +539,17 @@ int main(int argc, char *argv[]) {
 
     } //done time
 
+  //Close down any file handles we have
+  for(auto& f : nexus_outfiles){
+    f.second.close();
+  }
+  //std::cout<<"Clearning manger with "<<manager.use_count()<<" references\n";
+  //Need to release all formulation references, those stored in hy_features object and the
+  //references used by the formulation manager
+  //this will ensure all formulation destructors are called and resources released.
+  features.release_formulations();
+  manager->clear();
```

Comment:
Maybe I'm not understanding it, what's the reason for using shared pointers if we're going to handle manually decrementing the reference counts anyways? It would make more sense to put `features` and `manager` within an RAII block so they go out of scope before needing to call these.

Reply:
Jumping around here in my responses, but they are all somewhat related.

`Formulations` exist as shared pointers because we want to ensure we only create one formulation per feature and reuse that formulation in different contexts.

The formulation manager (which superseded the network indexing I'll discuss momentarily) exists entirely to abstract and encapsulate the parsing, configuring, and instantiation of formulations, i.e. ready to execute models associated with some feature.

It does this reasonalby well, and ultimately creates a `map` of `feature_id -> Formulation`.  Each formulation must be instantiated and ready to "run" before be we can begin using them in the modeling steps, so this container does just that.  The `hy_features` container, then, is responsible for aligning the topology and network with these formulations.  It uses the constructed formulations from the manager and "ties" them to the HY_Catchment realization.  The `network` uses feature indexing and maps to provide the indirection back to to the formulation when used in the actual computational steps...

I'm not sure if that helped in understanding or not.  I can try to articulate that more completely if needed.  To your point on using RAII blocks...yes, that would be one solution.  The functions to explicitly release the resources by clearing the underlying maps that hold the references are another.  Having these functions provides some additional flexibility beyond relying on the RAII blocks, though.  For example, a `Network` still has a valid feature index that may be useful/needed without having a need for the model formulation associated with the features indexed by the network (the complete HY_Catchment "realization").  Similarly, the manager actually manages a little bit more than *just* formulations -- it manages time and other configurations that have utility without require references to the model formuations for each catchment.